components/pool/PoolForm.tsx

```diff
@@ -0,0 +1,295 @@
+import React, { useEffect, useState } from "react";
+import Image from "next/image";
+import { Box, Button, TextField, Typography } from "@mui/material";
+import SwapHorizIcon from "@mui/icons-material/SwapHoriz";
+import { Heading, SmallText } from "../../theme/StyledComponents";
+import TokenPicker from "../../components/shared/TokenPicker";
+import { AssetInfo } from "../../providers/SupportedAssetsProvider";
+import { useWallet } from "../../providers/SupportedWalletProvider";
+import { PoolAction, usePool } from "../../providers/PoolProvider";
+import { Amount } from "../../utils/Amount";
+
+const ROUND_UP = 1;
+
+type PoolValues = {
+	poolAsset: number | string;
+	coreAsset: number | string;
+};
+
+const PoolForm: React.FC<{}> = () => {
+	const [poolAction, setPoolAction] = useState<string>(PoolAction.ADD);
+	const [poolAsset, setPoolAsset] = useState<AssetInfo>(null);
+	const [poolAssetAmount, setPoolAssetAmount] = useState<Amount>(new Amount(0));
+	const [coreAmount, setCoreAmount] = useState<Amount>(new Amount(0));
+	const [poolLiquidity, setPoolLiquidity] = useState<PoolValues>();
+	const [userBalances, setUserBalances] = useState<PoolValues>();
+	const { balances } = useWallet();
+	const {
+		coreAsset,
+		estimatedFee,
+		exchangePool,
+		updateExchangePool,
+		defineExtrinsic,
+		getUserPoolShare,
+		userPoolShare,
+		sendExtrinsic,
+	} = usePool();
+
+	//get user and pool balances
+	useEffect(() => {
+		if (!balances || !poolAsset || !coreAsset) return;
+		updateExchangePool(poolAsset);
+		getUserPoolShare(poolAsset);
+
+		const userPoolToken = balances.find(
+			(asset) => asset.symbol === poolAsset.symbol
+		);
+		const userCore = balances.find(
+			(asset) => asset.symbol === coreAsset.symbol
+		);
+
+		setUserBalances({
+			poolAsset: userPoolToken.value,
+			coreAsset: userCore.value,
+		});
+		//eslint-disable-next-line
+	}, [balances, poolAsset, coreAsset]);
+
+	//set core amount from token amount
+	//define extrinsic & estimate fee
+	useEffect(() => {
+		if (!exchangePool || !poolAsset || !poolAssetAmount) return;
+		if (poolAssetAmount.toNumber() <= 0) setCoreAmount(new Amount(0));
+		if (
+			exchangePool.coreAssetBalance.isZero() ||
+			exchangePool.assetBalance.isZero()
+		) {
+			setCoreAmount(poolAssetAmount);
+		} else {
+			const coreAmount = new Amount(poolAssetAmount)
+				.mul(exchangePool.coreAssetBalance)
+				.div(exchangePool.assetBalance)
+				.subn(ROUND_UP);
+
+			if (coreAmount.toNumber() <= 0) {
+				setCoreAmount(new Amount(0));
+			} else {
+				setCoreAmount(new Amount(coreAmount));
+				defineExtrinsic(poolAsset, poolAssetAmount, coreAmount, poolAction);
+			}
+		}
+		//eslint-disable-next-line
```

Comment:
Ditto

Reply:
'defineExtrinsic', 'exchangePool', and 'poolAction', i can safely add poolAction, otherwise infinite loop

I think this may be because these functions are wrapped in useCallback() ?

Reply:
Sure, please add `poolAction` in if you think it's safe.

Agreed, wrapped in useCallback causing the function itself changing over time thus cascading the effects down the tree. This generally a sign of something that are tangled together that should not be, but we can re-visit it later, add the `FIXME` comment in similar to my comment above and we are good to go.